src/traits.rs

```diff
@@ -183,6 +186,120 @@ impl<D: DistAlgorithm> From<TargetedMessage<D::Message, D::NodeId>> for Step<D>
     }
 }
 
+impl<D, I> From<I> for Step<D>
+where
+    D: DistAlgorithm,
+    I: IntoIterator<Item = TargetedMessage<D::Message, D::NodeId>>,
+{
+    fn from(msgs: I) -> Self {
+        Step {
+            messages: msgs.into_iter().collect(),
+            ..Step::default()
+        }
+    }
+}
+
+/// An interface to objects with epoch numbers. Different algorithms may have different internal
+/// notion of _epoch_. This interface summarizes the properties that are essential for the message
+/// sender queue.
+pub trait Epoched {
+    type Epoch: Copy + Message + Default + Eq + Ord + Serialize + DeserializeOwned;
+
+    /// Returns the object's epoch number.
+    fn epoch(&self) -> Self::Epoch;
+}
+
+impl<M: Epoched, N> Epoched for TargetedMessage<M, N> {
+    type Epoch = <M as Epoched>::Epoch;
+
+    fn epoch(&self) -> Self::Epoch {
+        self.message.epoch()
+    }
+}
+
+impl<'i, D> Step<D>
+where
+    D: DistAlgorithm,
+    <D as DistAlgorithm>::NodeId: NodeIdT + Rand,
+    <D as DistAlgorithm>::Message:
+        'i + Clone + SenderQueueableMessage + Serialize + DeserializeOwned,
+    <D as DistAlgorithm>::Output: Epoched,
+{
+    /// Removes and returns any messages that are not yet accepted by remote nodes according to the
+    /// mapping `remote_epochs`. This way the returned messages are postponed until later, and the
+    /// remaining messages can be sent to remote nodes without delay.
+    pub fn defer_messages(
+        &mut self,
+        peer_epochs: &'i BTreeMap<D::NodeId, <D::Message as Epoched>::Epoch>,
+        max_future_epochs: u64,
+    ) -> Vec<(D::NodeId, D::Message)>
+    where
+        <D as DistAlgorithm>::NodeId: 'i,
+    {
+        let messages = &mut self.messages;
+        let (mut passed_msgs, failed_msgs): (Vec<_>, Vec<_>) =
+            messages
+                .drain(..)
+                .partition(|TargetedMessage { target, message }| match target {
+                    Target::All => peer_epochs
+                        .values()
+                        .all(|&them| message.is_accepted(them, max_future_epochs)),
+                    Target::Node(id) => peer_epochs
+                        .get(&id)
+                        .map_or(false, |&them| message.is_accepted(them, max_future_epochs)),
+                });
```

Comment:
I still feel it would be easier to just iterate through all the messages in the big loop below, and populate `passed_msgs` there, too. That would avoid allocating and then consuming `failed_msgs`.
(Can also be refactored in another PR, of course.)